### PR TITLE
add signal to modify pressure scaling

### PR DESCRIPTION
**Pull request by tjhei**
FYI @quangx @gassmoeller

### CODE REVIEW

File: include/aspect/simulator_signals.h [hunk 1]
```diff
@@ -203,6 +203,18 @@ namespace aspect
...
+     * This signal is called whenever the pressure scaling is computed, see
+     * Simulator::compute_pressure_scaling_factor(), and allows inspection
+     * and/or modification of the computed factor.
+     *
+     * The argument @p pressure_scaling contains the computed pressure scaling (the ratio
+     * of the reference viscosity @p reference_viscosity computed by averaging
+     * the viscosity in the domain and the length scale @p length_scale reported
+     * by the geometry model) and can be used to modify the scaling factor.
+     */
+    boost::signals2::signal<void (double &pressure_scaling, const double reference_viscosity, const double length_scale)>  modify_pressure_scaling;
```
Comment (gassmoeller) [inline include/aspect/simulator_signals.h:216]: Did you look into making this:
```suggestion
    boost::signals2::signal<double (const double pressure_scaling, const double reference_viscosity, const double length_scale)>  modify_pressure_scaling;
```

and then below do:
```
if (signals.modify_pressure_scaling.empty() == false)
  result = signals.modify_pressure_scaling(result, reference_viscosity, length_scale);
```
As far as I understand the boost documentation if there are several connected slots the signal will by default return the return value of the last slot function.

**Review by gassmoeller — COMMENTED** [review]
I am ok with the change either this way, or the way I described in my comment. Let me know what you think about the suggestion.

File: include/aspect/simulator_signals.h (hunk 1 above)
Comment (bangerth) [reply include/aspect/simulator_signals.h:216]: I am in favor of @gassmoeller 's style.

**Comment by tjhei** [comment]
updated with the suggestion.

**Review by gassmoeller — APPROVED** [review]

File: source/simulator/helper_functions.cc [hunk 2]
```diff
@@ -461,8 +461,14 @@ namespace aspect
...
     const double reference_viscosity = std::exp(values[0]/values[1]);
+    const double length_scale = geometry_model->length_scale();
 
-    return reference_viscosity / geometry_model->length_scale();
+    // Allow the user to inspect and/or overwrite our result:
+    const double result = reference_viscosity / length_scale;
+    if (boost::optional<double> value = signals.modify_pressure_scaling(result, reference_viscosity, length_scale))
+      return *value;
+    else
+      return result;
```
Comment (bangerth) [inline source/simulator/helper_functions.cc:471]: You can combine this into
```suggestion
    return signals.modify_pressure_scaling(result, reference_viscosity, length_scale).value_or(result);
```
See https://www.boost.org/doc/libs/1_82_0/libs/optional/doc/html/boost_optional/reference/header__boost_optional_optional_hpp_/detailed_semantics___optional_values.html#reference_optional_value_or_move

Comment (bangerth) [reply source/simulator/helper_functions.cc:471]: Whether that's easier to read I don't know.

Comment (tjhei) [reply source/simulator/helper_functions.cc:471]: I don't care much either way. Let me know if you want me to make that change (PR is already merged).

**Pull request merged**